### PR TITLE
feat: ucan support

### LINKED ISSUE #1411: Add API support for UCAN.storage with service sdk

### PR DESCRIPTION
**Pull request by hugomrdias**
This PR add the following

- GET `/did` endpoint, returns service did
- POST `/user/did` endpoint, allows the user to register a DID in their account using normal API keys
- POST `/ucan/token`, enables a user to get a new ucan jwt signed by the service with full capabilities and ttl of 2 weeks. Supports refreshing ucan with ucan.
- `/upload` support ucan auth and saves ucan data to `meta` in the database

closes #1411

### CODE REVIEW

File: packages/api/src/constants.js [hunk 1]
```diff
@@ -1,6 +1,7 @@
 // let MAGIC_SECRET_KEY, SALT, PINATA_JWT, SENTRY_DSN, DATABASE_TOKEN, CLUSTER_SERVICE, LOGTAIL_TOKEN, MAILCHIMP_API_KEY, METAPLEX_AUTH_TOKEN
 
 export const secrets = {
+  privateKey: PRIVATE_KEY,
```
Comment (vasco-santos) [inline packages/api/src/constants.js:4]: Can you add to README that we need this wrangler secret?

File: packages/api/src/index.js [hunk 2]
```diff
@@ -97,6 +103,9 @@ r.add('delete', '/:cid', withMode(nftDelete, RW), [postCors])
 // Temporary Metaplex upload route, mapped to metaplex user account.
 r.add('post', '/metaplex/upload', withMode(metaplexUpload, RW), [postCors])
 
+// User
+r.add('post', '/user/did', withMode(userDid, RW), [postCors])
```
Comment (vasco-santos) [inline packages/api/src/index.js:107]: Can we rename handler to `didRegister`?

File: packages/api/src/index.js [hunk 3]
```diff
@@ -76,7 +79,10 @@ r.add(
 const psa = (handler, mode) =>
   withPsaErrorHandler(withPinningAuthorized(withMode(handler, mode)))
 
+r.add('get', '/did', withMode(did, RO), [postCors])
```
Comment (vasco-santos) [inline packages/api/src/index.js:82]: can we use `didGet` for consistency with other handlers namings?

File: packages/api/src/index.js [hunk 4]
```diff
@@ -76,7 +79,10 @@ r.add(
...
+
 // Login
+r.add('post', '/ucan/token', withMode(ucanToken, RW), [postCors])
```
Comment (vasco-santos) [inline packages/api/src/index.js:85]: can we rename handler to `ucanTokenGet`?

File: packages/api/src/index.js (hunk 3 above)
Comment (hugomrdias) [reply packages/api/src/index.js:82]: all the handlers follow the file name

Comment (vasco-santos) [reply packages/api/src/index.js:82]: yes, we can also rename the filename.

Comment (hugomrdias) [reply packages/api/src/index.js:82]: what im saying is that the current file names and function name are consistent with the other endpoint

File: packages/api/test/nfts-upload.spec.js [hunk 5]
```diff
@@ -302,4 +304,62 @@ describe('NFT Upload ', () => {
 
     assert.equal(backup.url, backupUrl)
   })
+
+  it.only('should upload a single file using ucan', async () => {
```
Comment (alanshaw) [inline packages/api/test/nfts-upload.spec.js:308]: Remove `.only`

**Review by alanshaw — COMMENTED** [review]
@yusefnapora are docs for this feature in the pipeline? If no can you open an issue?

File: packages/api/src/routes/user-did-register.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
+import { validate } from '../utils/auth.js'
+import { JSONResponse } from '../utils/json-response.js'
+import { parse } from 'ucan-storage/did'
+
+/** @type {import('../bindings').Handler} */
+export const userDidRegister = async (event, ctx) => {
```
Comment (flea89) [inline packages/api/src/routes/user-did-register.js:6]: nit
`userDidRegister` => `userDIDRegister` 
A bit more readable and avoid having `did` read like a verb.

**Review by flea89 — COMMENTED** [review]
A bit late to the party. But this looks great! 👏 
Super exciting to see UCAN land on nft.storage 🎉

File: README.md [hunk 7]
```diff
@@ -87,6 +87,7 @@ METAPLEX_AUTH_TOKEN=secret
 # needs to be real so create a personal magic.link account or use the staging key
 MAGIC_SECRET_KEY=secret
 LOGTAIL_TOKEN=secret
+PRIVATE_KEY=xmbtWjE9eYuAxae9G65lQSkw36HV6H+0LSFq2aKqVwY=
```
Comment (adamalton) [inline README.md:90]: I'm guessing this is some kind of dev/testing key(??), but I think a comment would be good to explain what this specific value is and maybe which public key it's linked to.

File: packages/api/test/did.spec.js [hunk 8]
```diff
@@ -0,0 +1,13 @@
+import assert from 'assert'
+
+describe('Service DID', () => {
+  it('should return the service DID', async () => {
+    const res = await fetch(`did`)
+    const { ok, value } = await res.json()
+
+    assert.equal(
+      value,
+      'did:key:z6MkgBRFJi7Ew2mzN7JtYN7nkq8M8qTHhLbsTEvV5j4ajWZS'
```
[the remark this answers is not in the document]
Comment (hugomrdias) [reply packages/api/test/did.spec.js:10]: this is the did:key format https://w3c-ccg.github.io/did-method-key/#format

File: packages/api/src/utils/auth.js [hunk 9]
```diff
@@ -3,17 +3,34 @@ import { secrets } from '../constants.js'
...
  * @param {import('../bindings').RouteContext} ctx
+ * @param {{checkUcan: boolean}} [options]
  */
-export async function validate(event, { log, db }) {
+export async function validate(event, { log, db, ucanService }, options) {
   const auth = event.request.headers.get('Authorization') || ''
   const token = magic.utils.parseAuthorizationHeader(auth)
 
+  if (options?.checkUcan && Ucan.isUcan(token)) {
+    const { root, cap } = await ucanService.validateFromCaps(token)
```
[the remark this answers is not in the document]
Comment (hugomrdias) [reply packages/api/src/utils/auth.js:20]: we dont know the root ucan at this point

**Pull request merged**